### PR TITLE
feat: added support for private registries on docker

### LINKED ISSUE #1342: Support private container registries
### Background & motivation

An occasional, but persistent, request we get is for supporting private container registries. We've been able to get by with only public registries so far, but we should have a way to support them.

This also ties in with handling secrets, as the secrets Kurtosis stores to access private repositories/registries need to be stored securely.

### Desired behaviour

Kurtosis has a way to access private container registries.

### How important is this to you?

Critical; Kurtosis is unusable for me without it.

### PR DESCRIPTION
**Pull request by h4ck3rk3y**
This adds support for private container registries

TODO
- [x] add docs
- [x] cleanup all the nils and maybe make the FetchImage interface nicer
- [x] perhaps add e2e tests (skipped)

Closes #1342

### CODE REVIEW

File: core/server/api_container/server/startosis_engine/startosis_validator/validator_environment.go [hunk 1]
```diff
@@ -60,6 +63,10 @@ func (environment *ValidatorEnvironment) AppendRequiredImageBuild(containerImage
 	environment.imagesToBuild[containerImage] = imageBuildSpec
 }
 
+func (environmemt *ValidatorEnvironment) AppendImageToPullWithAuth(containerImage string, registrySpec *image_registry_spec.ImageRegistrySpec) {
+	environmemt.imagesToPullWithAuth[containerImage] = registrySpec
+}
+
 func (environment *ValidatorEnvironment) GetNumberOfContainerImagesToProcess() uint32 {
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/startosis_validator/validator_environment.go:68]: think you've gotta update this to account for `environment.imagesToPullWithAuth`

File: core/server/api_container/server/startosis_engine/startosis_validator/validator_environment.go [hunk 2]
```diff
@@ -14,6 +15,7 @@ import (
 type ValidatorEnvironment struct {
 	imagesToPull                  map[string]bool // "set" of images that need to be downloaded
 	imagesToBuild                 map[string]*image_build_spec.ImageBuildSpec
+	imagesToPullWithAuth          map[string]*image_registry_spec.ImageRegistrySpec
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/startosis_validator/validator_environment.go]: wonder if it's possible to merge `imagesToPull` and `imagesToPullWithAuth` into one `imagesToPull map[string]*image_registry_spec.ImageRegistrySpec`, where raw container image strings just have nil `ImageRegistrySpec`/ no auth/default registry? wdyt?

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_registry_spec.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return &ImageRegistrySpec{
+		KurtosisValueTypeDefault: copiedValueType,
+	}, nil
+}
+
+// GetImage returns the image that needs to be pulled
+func (irs *ImageRegistrySpec) GetImage() (string, *startosis_errors.InterpretationError) {
```
Comment (tedim52) [inline core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_registry_spec.go:90]: maybe suggest `GetImageName` to be consistent with `ImageBuildSpec` but up to you

**Review by tedim52 — APPROVED** [review]
looks good to me

File: core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_registry_spec.go (hunk 3 above)
Comment (tedim52) [reply core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_registry_spec.go:90]: irs - internal revenue service....

File: core/server/api_container/server/startosis_engine/startosis_validator/validator_environment.go (hunk 1 above)
Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/startosis_validator/validator_environment.go:68]: good idea

File: core/server/api_container/server/startosis_engine/startosis_validator/validator_environment.go (hunk 2 above)
Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/startosis_validator/validator_environment.go]: Yep I thought of this as well but was worried about introducing nulls, though it should be okay.

**Pull request merged**